### ISSUE TITLE
`explicit_auto_deref` gives wrong suggestions on `Box`+trait

### ISSUE THREAD

**Issue by greenhandatsjtu** [rust-lang/rust-clippy#9165]
### Summary

When I want to get the reference of a struct which implements a trait from `Box`, clippy will give wrong suggestions about `explicit_auto_deref`

### Lint Name

explicit_auto_deref

### Reproducer

Today build workflow of my repository failed (view details via: https://github.com/greenhandatsjtu/casbin-rs/runs/7314989421?check_suite_focus=true), I found it's cargo clippy that failed and it gave several suggestions on `explicit_auto_deref`.

I tried this code:

```rust
pub struct Enforcer {
    model: Box<dyn Model>,
    // ...
}

pub trait Model: Send + Sync {
    // ...
}

#[async_trait]
impl CoreApi for Enforcer {
    // ...
    #[inline]
    fn get_model(&self) -> &dyn Model {
        &*self.model
    }
}
```

I saw this happen:

```
error: deref which would be done by auto-deref
   --> src/enforcer.rs:281:10
    |
281 |         &*self.model
    |          ^^^^^^^^^^^ help: try this: `self.model`
    |
    = note: `-D clippy::explicit-auto-deref` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
```
I changed to `&self.model` as suggestion but got this error:
```
error[E0277]: the trait bound `std::boxed::Box<(dyn model::Model + 'static)>: model::Model` is not satisfied
   --> src/enforcer.rs:281:9
    |
281 |         &self.model
    |         ^^^^^^^^^^^ the trait `model::Model` is not implemented for `std::boxed::Box<(dyn model::Model + 'static)>`
    |
    = help: the trait `model::Model` is implemented for `model::default_model::DefaultModel`
    = note: required for the cast from `std::boxed::Box<(dyn model::Model + 'static)>` to the object type `dyn model::Model`

For more information about this error, try `rustc --explain E0277`.
```
I expected to see this happen:
No error

### Version

```text
cargo 1.64.0-nightly (b1dd22e66 2022-07-09)
release: 1.64.0-nightly
commit-hash: b1dd22e668af5279e13a071ad4b17435bd6bfa4c
commit-date: 2022-07-09
host: x86_64-unknown-linux-gnu
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.83.1-DEV (sys:0.4.55+curl-7.83.1 vendored ssl:OpenSSL/1.1.1n)
os: Manjaro 21.3.2 (Ruah) [64-bit]
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by greenhandatsjtu** [comment]
> Same as #9143

Oh I see, thank you!
